### PR TITLE
Provisioning: datasource and dashboard org by name

### LINKED ISSUE #77282: Provisioning: datasource and dashboard org by name
**Why is this needed**:

We are using multiple Orgs. Each org has a loki and a prometheus datasource. We also populate dashboards specific to some Orgs.

Currently we need to keep a map of Org names to Org IDs, to properly [provision](https://grafana.com/docs/grafana/latest/administration/provisioning/) datasources and dashboards. We are using somethink like this:

```yaml
# datasources.yaml.j2 (Jinja2)
apiVersion: 1
datasources:
  {% for org in orgs %}
  - name: Loki {{ org.name }}
    orgId:  {{ org.id }}  # <===== HERE
    type: loki
    access: proxy
    url: http://<REDACTED>
    jsonData:
        httpHeaderName1: X-Scope-OrgID
    secureJsonData:
        httpHeaderValue1: {{ org.name }}
  {% endfor %}
```

**What would you like to be added**:

Allow to use Org **name** in datasources and dashboards provisioned.

```yaml
# datasources.yaml.j2 (Jinja2)
apiVersion: 1
datasources:
  {% for org in orgs %}
  - name: Loki {{ org.name }}
    orgName:  {{ org.name }}  # <===== HERE
    type: loki
    access: proxy
    url: http://<REDACTED>
    jsonData:
        httpHeaderName1: X-Scope-OrgID
    secureJsonData:
        httpHeaderValue1: {{ org.name }}
  {% endfor %}
```

This is done in https://github.com/grafana/grafana/pull/76937.

**Who is this feature for?**

Users of [file provisioning](https://grafana.com/docs/grafana/latest/administration/provisioning/) having multiple Orgs.

### PR DESCRIPTION
**Pull request by sathieu**
**What is this feature?**

Improve provisioning to set org for datasources and dashboards by name.

**Why do we need this feature?**

I don't want to remember the Org Ids. 

**Who is this feature for?**

Every user using provisioning with multiple Orgs

**Which issue(s) does this PR fix?**:

Fixes #77282

**Special notes for your reviewer:**

There are two commits (one for datasource, one for dashboard). If needed I can split in two PRs.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by andresmgot** [comment]
hi @sathieu, could you create a GitHub issues with more details about the problem and associate this PR with that? That will help with the review. Thanks.

**Comment by sathieu** [comment]
Hi @andresmgot Of course! I've created https://github.com/grafana/grafana/issues/77282.

File: pkg/services/org/orgtest/fake.go
Comment (andresmgot) [inline pkg/services/org/orgtest/fake.go:26]: Shouldn't this be 0?

```suggestion
	return &FakeOrgService{ExpectedOrg: &org.Org{ID: 0}}
```

**Review by andresmgot — COMMENTED** [review]
This sounds good to me! but someone from @grafana/backend-platform should take a look at it.

File: pkg/services/org/orgtest/fake.go
Comment (sathieu) [reply pkg/services/org/orgtest/fake.go:26]: @andresmgot No. This Org is tested against in https://github.com/grafana/grafana/pull/76937/files#diff-8f8b2e7e8794fdda540742ab01df72e23bcb091b6bbe05a0ec33f98888d2c9a1R119.

This `FakeOrg` struct can be changed to an in-memory implementation. This would improve tests. I can do this (maybe in another PR?) if you want.

**Comment by sathieu** [comment]
@andresmgot What is the status of this PR? It looks like no one from @grafana/backend-platform has received the ping. Maybe @gamab ?

**Comment by andresmgot** [comment]
Hi @sathieu, @mildwonkey from the @grafana/backend-platform replied in the issue: https://github.com/grafana/grafana/issues/77282#issuecomment-1810721909

It seems there are concerns about this approach. You can keep the conversation there.

**Comment by sathieu** [comment]
Not stale.

This is waiting for org uuid transition, or merged before as similar code is already present.

**Comment by sathieu** [comment]
Not stale. We currently use a patched version to have this (simple) feature.

**Comment by sathieu** [comment]
(resolved conflicts)

**Comment by sathieu** [comment]
Not stale

**Review by lwandz13 — APPROVED** [review]
lgtm!

**Comment by sathieu** [comment]
not stale ...

**Comment by sathieu** [comment]
not stale

**Comment by sathieu** [comment]
Hello,

We ended up writing our Kubernetes Operator to manage:

- GrafanaInstances (i.e. connection to API)
- GrafanaOrgs
- GrafaOrgDatasources
- GrafanaOrgDashboards

It' available at: https://gitlab.com/kubitus-project/grafana-org-operator/#grafana-org-operator. Feedback welcome!

**Pull request closed without merging**
